### PR TITLE
Stream input and output support for optional collections

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR adds support for reading and writing optional collections via
`StreamInput` and `StreamOutput`.

A use case for this came up via https://github.com/elastic/elasticsearch/pull/87924#discussion_r907948009.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:27]: nit: is this import strictly necessary? the existing qualified Writeable.Writer is more symmetric with Readable.Writer

File: server/src/test/java/org/elasticsearch/common/io/stream/AbstractStreamTests.java
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/common/io/stream/AbstractStreamTests.java:272]: Instead of relying on randomization, can we have explicit tests for the relevant cases? The data for the collection can be randomized, but we should explicitly test null and not null.

**Review by rjernst — COMMENTED** [review]
A couple thoughts

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java
Comment (n1v0lg) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:27]: good catch, will revert 👍

**Review by ywangd — APPROVED** [review]
LGTM pending green CI

I'd also wait for @rjernst's approval. Thanks!

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**
